File: src/main/java/edu/cornell/cs/apl/viaduct/imp/ast/Metadata.java
package edu.cornell.cs.apl.viaduct.imp.ast;

import java.util.Objects;
import javax.annotation.Nullable;

/**
 * A wrapper for metadata information. The main use of this class is getting AutoValue to ignore
 * certain fields in the definitions of {@link Object#equals(Object)} and {@link Object#hashCode()}.
 *
 * <p>All instances of this class compare equal to each other.
 *
 * @param <T> type of wrapped values
 */
final class Metadata<T> {
  private final T data;

  Metadata(@Nullable T data) {
    this.data = data;
  }

  @Nullable
  T getData() {
    return data;
  }

  @Override
  public boolean equals(Object that) {
    return that instanceof Metadata;
  }

  @Override
  public int hashCode() {
    return 0;
  }

  @Override
  public String toString() {
    return Objects.toString(getData());
  }
}

File: src/main/kotlin/edu/cornell/cs/apl/viaduct/imp/ast2/Metadata.kt
package edu.cornell.cs.apl.viaduct.imp.ast2

/**
 * A wrapper for metadata information. Used to mark properties that should be ignored when checking
 * equality.
 *
 * All instances of this class compare equal to each other.
 *
 * @param T type of wrapped values
 */
class Metadata<T>(var data: T? = null) {
    override operator fun equals(other: Any?): Boolean {
        return other is Metadata<*>
    }

    override fun hashCode(): Int {
        return 0
    }

    override fun toString(): String {
        return data.toString()
    }
}
